File: androidApp/src/main/java/dev/haroldjose/familysharedlist/android/presentationLayer/pages/familyList/views/components/rowItem/FamilyListRowItemWithProduct.kt
package dev.haroldjose.familysharedlist.android.presentationLayer.pages.familyList.views.components.rowItem

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.BottomSheetScaffoldState
import androidx.compose.material3.Card
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.rememberBottomSheetScaffoldState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import dev.haroldjose.familysharedlist.android.app.MyApplicationTheme
import dev.haroldjose.familysharedlist.android.presentationLayer.pages.familyList.viewmodels.FamilyListViewModelMocked
import dev.haroldjose.familysharedlist.android.presentationLayer.pages.familyList.viewmodels.IFamilyListViewModel
import dev.haroldjose.familysharedlist.domainLayer.extensions.Samples
import dev.haroldjose.familysharedlist.domainLayer.models.FamilyListModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FamilyListRowItemWithProduct(
    item: FamilyListModel,
    viewModel: IFamilyListViewModel,
    bottomSheetScaffoldState: BottomSheetScaffoldState,
) {
    val coroutineScope = rememberCoroutineScope()
    val itemMutable = remember { mutableStateOf(item) }
    val nameInEditMode = remember { mutableStateOf(false) }
    val priceInEditMode = remember { mutableStateOf(false) }

    Card(
        modifier = Modifier.fillMaxWidth(),
        shape = RoundedCornerShape(10.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            when {
                nameInEditMode.value -> {
                    ColumnRigthEditingName(
                        itemMutable,
                        nameInEditMode,
                        coroutineScope,
                        viewModel
                    )
                }

                priceInEditMode.value -> {
                    ColumnRigthEditingPrice(
                        itemMutable,
                        priceInEditMode,
                        coroutineScope,
                        viewModel
                    )
                }

                else -> {
                    ColumnLeftDefault(
                        itemMutable,
                        viewModel
                    )
                    ColumnRigthDefault(
                        itemMutable,
                        nameInEditMode,
                        priceInEditMode,
                        coroutineScope,
                        viewModel,
                        bottomSheetScaffoldState
                    )
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Preview
@Composable
fun FamilyListRowItemWithProduct_Preview() {
    val bottomSheetScaffoldState = rememberBottomSheetScaffoldState()
    val item = Samples.FamilyList.nutella
    MyApplicationTheme {
        FamilyListRowItemWithProduct(
            item = item ,
            viewModel = FamilyListViewModelMocked(),
            bottomSheetScaffoldState = bottomSheetScaffoldState
        )
    }
}

File: iosApp/iosApp/presentationLayer/pages/familyList/views/components/rowItem/FamilyListRowItemWithProduct.swift
import SwiftUI
import shared

struct FamilyListRowItemWithProduct<ViewModel>: View where ViewModel: FamilyListViewModelProtocol {

    var viewModel: ViewModel
    @State var item: FamilyListModel

    @State private var nameTextFieldValue = ""
    @State private var nameInEditMode = false
    @State private var priceTextFieldValue: Int = 0
    @State private var priceInEditMode = false
    @State private var isBusy = false

    var body: some View {
        HStack {
            switch (nameInEditMode,priceInEditMode) {
            case (true,_):
                ColumnRigthEditingName(
                    item: $item, 
                    nameTextFieldValue: $nameTextFieldValue,
                    nameInEditMode: $nameInEditMode,
                    viewModel: viewModel,
                    isBusy: $isBusy
                )
            case (_,true):
                ColumnRigthEditingPrice(
                    item: $item, 
                    priceTextFieldValue: $priceTextFieldValue,
                    priceInEditMode: $priceInEditMode,
                    viewModel: viewModel,
                    isBusy: $isBusy
                )
            default:
                ColumnLeftDefault(
                    item: $item,
                    viewModel: viewModel
                )
                ColumnRigthDefault(
                    item: $item,
                    nameTextFieldValue: $nameTextFieldValue,
                    nameInEditMode: $nameInEditMode,
                    priceTextFieldValue: $priceTextFieldValue,
                    priceInEditMode: $priceInEditMode,
                    viewModel: viewModel,
                    isBusy: $isBusy
                )
            }
        }
    }
}

#Preview {
    FamilyListRowItemWithProduct(
        viewModel: FamilyListViewModelMocked(), 
        item: Samples.FamilyList.companion.nutella
    )
}
